battlesnake-rs/src/mcts_snake.rs

```diff
@@ -0,0 +1,442 @@
+                    borrowed.expand(&arena);
+
+                    borrowed.next_leaf_node(total_number_of_iterations)
+                } else {
+                    next_leaf_node
+                }
+            };
+
+            //Now we do a simulation for this leaf node
+            let score = next_leaf_node.simulate(&mut rng);
+
+            //We now need to backpropagate the score
+            next_leaf_node.backpropagate(score);
+        }
+
+        // We are outside the loop now and need to pick the best move
+        current_span.record("total_number_of_iterations", &total_number_of_iterations);
+        current_span.record(
+            "total_score",
+            &root_node.total_score.load(Ordering::Relaxed),
+        );
+
+        let best_child = root_node.best_child(root_node.number_of_visits.load(Ordering::Relaxed));
+        let chosen_move = best_child
+            .expect("The root should have a child")
+            .tree_context
+            .as_ref()
+            .expect("We found the best child of the root node, so it _should_ have a tree_context")
+            .r#move;
+
+        Ok(MoveOutput {
+            r#move: format!("{}", chosen_move),
+            shout: None,
+        })
+    }
+
+    fn end(&self) {
+        info!("Mcts has ended");
+    }
+}
+
+#[derive(Debug)]
+struct TreeContext<'arena, T> {
+    parent: RefCell<&'arena Node<'arena, T>>,
+    r#move: Move,
+}
+
+#[derive(Debug)]
+struct Node<'arena, T> {
+    game_state: T,
+    total_score: AtomicF64,
+    number_of_visits: AtomicUsize,
+    children: RefCell<Option<Vec<&'arena Node<'arena, T>>>>,
+    tree_context: Option<TreeContext<'arena, T>>,
+}
+
+#[derive(Debug)]
+pub struct Instrument {}
+impl SimulatorInstruments for Instrument {
+    fn observe_simulation(&self, _duration: std::time::Duration) {
+        //No-oping here
+    }
+}
+
+impl<'arena, T> Node<'arena, T> {
+    fn new(game_state: T) -> Self {
+        Self {
+            game_state,
+            total_score: AtomicF64::new(0.0),
+            number_of_visits: AtomicUsize::new(0),
+            children: RefCell::new(None),
+            tree_context: None,
+        }
+    }
+
+    fn new_with_parent(game_state: T, parent: &'arena Self, r#move: Move) -> Self {
+        Self {
+            game_state,
+            total_score: AtomicF64::new(0.0),
+            number_of_visits: AtomicUsize::new(0),
+            children: RefCell::new(None),
+            tree_context: Some(TreeContext {
+                parent: RefCell::new(parent),
+                r#move,
+            }),
+        }
+    }
+}
+
+impl<'arena, T> Node<'arena, T>
+where
+    T: SimulableGame<Instrument, 4>
+        + SnakeIDGettableGame
+        + RandomReasonableMovesGame
+        + Clone
+        + VictorDeterminableGame
+        + YouDeterminableGame,
+{
+    fn simulate(&self, rng: &mut ThreadRng) -> f64 {
+        // TODO: This clone might not be the best
+        let mut current_state = self.game_state.clone();
```

Comment:
I think we can use a `Cow` here. We won't actually need to write to this, but it will allow us to hold _either_ a reference or an object.

Here we can use a reference, and then below when we have the struct and can't reference it or the object will be freed. `Cow` will let us hold either of these 🎉